### PR TITLE
#4372 [FEATURE] Remove entity

### PR DESCRIPTION
**Pull request by Dufgui**
add remove option to entity sub generator

### CODE REVIEW

**Comment by PierreBesson** [comment]
:+1: If sufficiently tested I would be in favor of this feature. I can't see any drawbacks to it and in the case it breaks the project people will be saved by git.

File: generators/entity/files.js [hunk 1]
```diff
@@ -16,9 +17,142 @@ module.exports = {
...
+            this.fs.delete(this.destinationPath(this.jhipsterConfigDirectory + '/' + this.entityNameCapitalized + '.json'));
+        },
+
+        deleteDbFiles: function () {
+            if(!this.remove) return;
+
+            if (this.skipServer) return;
+
+            if (this.databaseType === 'sql') {
+                this.warning('Please update manually the liquibase to deal with data before drop the table');
```
Comment (deepu105) [inline generators/entity/files.js]: Why is this required. I think we can delete liquibase changelog files and remove the entry from master xml as well and then print a warning to ask user to update the DB manually or by running liquibase:diff goal after removing the entity

File: generators/entity/files.js [hunk 2]
```diff
@@ -16,9 +17,142 @@ module.exports = {
...
+            }
+            if (this.databaseType === 'cassandra') {
+                this.warning('Please update manually the cql script to deal with data before drop the column family');
```
Comment (deepu105) [inline generators/entity/files.js]: Messages needs to be corrected for grammer

File: generators/entity/index.js [hunk 3]
```diff
@@ -69,6 +69,14 @@ module.exports = EntityGenerator.extend({
             type: Boolean,
             defaults: false
         });
+
+        // This method adds support for a `--[no-]regenerate` flag
```
Comment (deepu105) [inline generators/entity/index.js]: Correct the comment

File: generators/generator-base.js [hunk 4]
```diff
@@ -163,12 +190,32 @@ Generator.prototype.addAdminElementTranslationKey = function (key, value, langua
 Generator.prototype.addEntityTranslationKey = function (key, value, language) {
     var fullPath = CLIENT_MAIN_SRC_DIR + 'i18n/' + language + '/global.json';
     try {
-        jhipsterUtils.rewriteFile({
+        jhipsterUtils.replaceContent({
```
Comment (deepu105) [inline generators/generator-base.js]: This method shouldn't be updated

**Comment by pascalgrimaud** [comment]
I agree it's a good idea and can help people

But my personnal opinion:
- I'm afraid it will lead into an increasing number of issues, if not well tested, and there are so much use cases...
- as the sub generator entity is already complex, I'd rather prefer it in another sub generator, but it's just me

**Comment by Dufgui** [comment]
@pascalgrimaud in fact, yes i can move my code to a new generator but to be maintained and coherent with the entity generator, i think it's a good idea to have all my code in the same file where the entity files is created.

**Comment by Dufgui** [comment]
So I add a unit test and fix a little bug on remove the entity json file

**Comment by jdubois** [comment]
Per our guidelines, I'd like first to have a ticket with a discussion, before implementing such a feature.

Then, I can already share my thoughts:

Pros:
- A lot of people are requesting this

Cons:
- I don't think it's useful, you should better use Git to rollback your changes
- It adds a lot more code in the generator, and that's going to be annoying to maintain (now we must always write code to "undo" what we do in the generator)

So I'm against such a feature, but I'm not the only one deciding here, and I know a lot of people want this, so we must discuss this.

**Comment by cbornet** [comment]
@jdubois I'm not sure using git to revert an existing entity is always possible: if you create Foo with JH 3.0 then upgrade it later with JH 3.10 then file creation can be scattered across multiple commits some of which will also probably be shared with other entities (the upgrade one). So you cannot just find the commit and revert it.
Now I understand the feeling about maintenance. My impression is that it is easy to maintain the list of files to delete and harder to maintain the files to modify (by regexp matching). So we could just provide the file remover and ask the user to do a `yo jhipster --with-entities` to clean the modified files.
Another idea would be to use the same idea as for `jhipster:upgrade` : in another branch, generate the app with the entity, commit, generate the app without, commit, then merge on the main branch.

**Comment by Dufgui** [comment]
it take 2 hour to doing it and 3h with the unit test, and some dev of my team asked it, so i propose it. Sorry to not create a issue for it.

For your first point:
I am agree with @cbornet , it 's complicated to use git for this because all our entity is created one times by jhipster-uml so to rollback it, we have one commit for all of them....

For your second point:
If you want i can split it in another sub generator, it's quite simple. I do it in the same because I think it's simplier for the evolution and keep it coherent with the entity generator.

**Comment by jdubois** [comment]
OK, @cbornet opinion as very important here, as he's using this for real (which is not my case!), so he knows better than me.

**Comment by Dufgui** [comment]
So I create the issue for the discussion.

**Comment by pascalgrimaud** [comment]
@Dufgui : FYI, you have changed your local git configuration, so in the commit logs https://github.com/jhipster/generator-jhipster/pull/4369/commits Github doesn't know you

**Comment by deepu105** [comment]
If we are doing this its better to be here than another sub gen

Thanks & regards,
Deepu

**Comment by deepu105** [comment]
I agree with the maintenance issue Julien highlighted. We already have a
solution to reduce it. Since im working on refactoring the write blocks to
use metadata rather than instructions it would be very easy to write or
delete files using a simple function so lot less code to maintain.
But then I agree that not many would use this.

Thanks & regards,
Deepu

**Comment by deepu105** [comment]
Yes you are right. This is a tough thing to get right. Especially
relationships as it cannot be easily removed. Then we can only help so
much, its better than nothing

Thanks & regards,
Deepu

**Comment by Dufgui** [comment]
So i close this one in favour of #4403

**Comment by Chris533** [comment]
In version 4+, should be mark that how to remove entity even by manual.

**Comment by Chris533** [comment]
Remove codes from files:
src\main\java\.\config\CacheConfiguration.java
src\main\resources\config\liquibase\master.xml
src\main\webapp\app\entities\entity.module.ts
src\main\webapp\app\layouts\navbar\navbar.component.html
src\main\webapp\i18n\en\global.json

Remove files from folders:
src\main\java\.\domain\*.java
src\main\java\.\repository\*Repository.java
src\main\java\.\web\rest\*Resource.java
src\main\webapp\app\entities\*
src\main\webapp\i18n\.\*.json
src\test\gatling\user-files\simulations\*GatlingTest.scala
src\test\java\jshop\bunlu\com\web\rest\*ResourceIntTest.java
src\test\javascript\e2e\entities\*.spec.ts
src\test\javascript\spec\app\entities\*

Otherwise, remove table from database, remove relations from where you used.

**Comment by gmarziou** [comment]
Or use git as advised previously and generate only on branches.
